### PR TITLE
[Profiler] Use libdatadog as buffer

### PR DESCRIPTION
**Pull request by kevingosse**
## Summary of changes

Instead of having an intermediary buffer in the `SamplesAggregator`, have the `SamplesCollector` directly push the samples to libdatadog.

## Reason for change

Slight optimization (since libdatadog directly aggregates the samples), plus laying the groundwork for supporting endpoint aggregation.

### CODE REVIEW

File: profiler/test/Datadog.Profiler.Native.Tests/SamplesAggregatorTest.cpp [hunk 1]
```diff
@@ -1,231 +1,84 @@
...
-    EXPECT_CALL(mockCollector, GetSamples()).Times(1).WillOnce(Return(ByMove(std::move(samples))));
-
-    auto aggregator = SamplesAggregator(&mockConfiguration, &threadsCpuManagerHelper, &mockExporter, &metricsSender, collector.get());
-
-    aggregator.Start();
-    std::this_thread::sleep_for(100ms);
-    aggregator.Stop();
-}
\ No newline at end of file
+//// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
```
Comment (gleocadie) [inline profiler/test/Datadog.Profiler.Native.Tests/SamplesAggregatorTest.cpp:1]: maybe remove this file no?

**Review by gleocadie — APPROVED** [review]
LGTM + one question

**Review by gleocadie — APPROVED** [review]
LGTM

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/LibddprofExporter.cpp [hunk 2]
```diff
@@ -312,30 +324,56 @@ void LibddprofExporter::Add(Sample const& sample)
...
+            scope.profileInfo.profile = nullptr;
+            scope.profileInfo.samplesCount = 0;
+        }
+
         const auto& applicationInfo = _applicationStore->GetApplicationInfo(std::string(runtimeId));
 
-        if (samplesCount <= 0)
+        if (profile == nullptr || samplesCount == 0)
         {
             Log::Debug("The profiler for application ", applicationInfo.ServiceName, " (runtime id:", runtimeId, ") have empty profile. Nothing will be send.");
```
Comment (chrisnas) [inline profiler/src/ProfilerEngine/Datadog.Profiler.Native/LibddprofExporter.cpp:372]:
```suggestion
            Log::Debug("The profiler for application ", applicationInfo.ServiceName, " (runtime id:", runtimeId, ") have empty profile. Nothing will be sent.");
```

Comment (kevingosse) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/LibddprofExporter.cpp]: Done in https://github.com/DataDog/dd-trace-dotnet/pull/2990/commits/f6491935e793d28a61b87bdd18c6e5175fbce622

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/SamplesCollector.cpp [hunk 3]
```diff
@@ -62,14 +67,67 @@ void SamplesCollector::Work()
     }
 }
 
-void SamplesCollector::CollectSamples()
+void SamplesCollector::ExportWork()
 {
-    for (auto const& samplesProvider : _samplesProviders)
+    const auto future = _exporterThreadPromise.get_future();
```
[the remark this answers is not in the document]
Comment (kevingosse) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/SamplesCollector.cpp:81]: Done in https://github.com/DataDog/dd-trace-dotnet/pull/2990/commits/f6491935e793d28a61b87bdd18c6e5175fbce622

File: profiler/src/ProfilerEngine/Datadog.Profiler.Native/SamplesCollector.cpp [hunk 4]
```diff
@@ -41,7 +43,10 @@ bool SamplesCollector::Stop()
     }
 
     _mustStop = true;
-    _worker.join();
+    _workerThread.join();
+
+    _exporterThreadPromise.set_value();
```
[the remark this answers is not in the document]
Comment (kevingosse) [reply profiler/src/ProfilerEngine/Datadog.Profiler.Native/SamplesCollector.cpp:50]: Done in https://github.com/DataDog/dd-trace-dotnet/pull/2990/commits/f6491935e793d28a61b87bdd18c6e5175fbce622

**Pull request merged**